### PR TITLE
Multi invest bug

### LINKED ISSUE #40: [BUG] A multi-period investment model allows generation from unavailable capacity
When run with multiple investment periods, the model will produce energy from units that are not yet built (see figure below). This leads to an unphysical amount of energy being produced in each investment period.

This issue may be closed when there is a pull request that resolves the issue or introduces a work-around, such as myopic investment.

Image

### PR DESCRIPTION
**Pull request by samgdotson**
This PR fixes the multi-investment issue where PyPSA was using generators from future years in earlier ones. Closes #40. 

## Why should we use pull requests to collaborate?
1. Pull requests serve as a built-in QA/QC step when developing research code and software.
2. The discussions within issues and pull requests document the rationale for certain decisions for future use.
3. Critically reviewing other people's code (and having your own code critically reviewed) is a great way to improve coding skills.
4. Pull requests don't *have* to be reviewed by someone working on the same project. Thus, PRs allow greater cross-program collaboration and break down silos.
5. Similarly, PRs allow interested third parties to meaningfully contribute to open-source projects. In the case of UCS, that may include enthusiastic members of the science network.
6. PRs protect your data and repository by preventing force pushes. In many cases, at least one other person must review the code before it is "merged" with the main code base.

## Where to focus attention
The following files should be prioritized for review: 

- [x] All files in the `scripts/` folder (these should be python source files, 8 in total).
- [x] All files in the `utils/` folder (these should be python source files, 2 in total).
- [x] The `Snakefile` file
- [x] The `config.yml` file

The jupyter notebook files can be ignored since they are experimental and will be removed in a future pull request.

## How to review a pull request

[This post](https://learn.microsoft.com/en-us/contribute/content/how-to-review-pull-request) goes over the steps for reviewing a pull request.

### Recommended checklist for reviewing a pull request

The guide below is modified from the [Advanced Reactors and Fuel Cycles (ARFC)](https://arfc.github.io/manual/guides/pull_requests) research group at the University of Illinois Urbana-Champaign

- [x] Read the PR description
- [ ] Read through all the changes, considering the following questions.
  - [ ] Is there anything you can praise about this PR? Start with
        praise.
  - [ ] Are variable names brief but descriptive?
  - [ ] Are new/changed functions no longer than a paragraph?
  - [ ] Do all function parameters have default values where appropriate?
  - [ ] Is the code clear and clean? (see Robert C. Martin's
        [Clean Code](https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29)
    - [ ] If something is unclear, ask for clarification
    - [ ] Can you offer a suggestion to make the functionality clearer?
  - [ ] Is there enough documentation?
  - [ ] Does the programming style meet the requirements of the
        repository ([PEP8](https://www.python.org/dev/peps/pep-0008/) for python,
        [google C++ style guide](https://google.github.io/styleguide/cppguide.html), etc.)
  - [ ] If a new feature has been added, or a bug fixed, has a test been
        added to confirm good behavior?
  - [ ] Does the test actually test the new/changed functionality?
  - [ ] Does the test successfully test edge cases?
  - [ ] Does the test successfully test corner cases?
  - [ ] If the repository has continuous integration: does the PR pass
        the tests?
  - [ ] If there is no continuous integration, check out the branch
        locally, build, and run the tests.
  - [ ] Do the tests pass on your machine?
  - [ ] Is the PR free of random cruft (built files, `.swp` files,
        etc.)?
  - [ ] Do all files in the PR belong in the repository?
  - [ ] If the PR deletes files, is this appropriate?
  - [ ] If the PR adds files or data, are these new files compatible
        with the repository license?
  - [ ] Make a review, leaving kind comments and suggesting changes
        where needed (to resolve the above).
  - [ ] Has the author resolved all of the comments and suggestions
        in your review?
- [ ] When you approve of the PR, merge and close it.
- [ ] Does this PR close an issue? If so, be sure to descriptively
      close this issue when the PR is merged
- [ ] Thank the author for th

### CODE REVIEW

File: scripts/add_electricity.py [hunk 1]
```diff
@@ -386,6 +394,93 @@ def attach_storage(
...
+        carriers = capacity_limits_df.columns
+        
+        df = capacity_limits_df.copy()
+        
+        for carrier in carriers:
+            for year in df.index:
+                try:
+                    capacity_lim = df.loc[year, carrier]
+                    energy_lim = capacity_lim*(8760/resolution)
+                    name = f'{carrier} Energy Limit 2025'
```
Comment (lshaver) [inline scripts/add_electricity.py:470]: Shouldn't year be a variable year, rather than fixed at 2025? Per the docstring this function applies limits for a given year, not a fixed year.

File: scripts/add_electricity.py [hunk 2]
```diff
@@ -386,6 +394,93 @@ def attach_storage(
...
+                    n.add(class_name="GlobalConstraint",
+                        type="operational_limit",
+                        investment_period=2025,
```
Comment (samgdotson) [inline scripts/add_electricity.py:473]:
```suggestion
                        investment_period=year,
```

**Review by lshaver — APPROVED** [review]
With the suggested changes I approve this PR. Lots of good work here!
